xml.mod/glue.c

```diff
@@ -259,6 +261,95 @@ BBString * bmx_mxmlElementGetAttrByIndex(mxml_node_t * node, int index, BBString
 	}
 }
 
+BBString * bmx_mxmlElementGetAttrByIndexNoName(mxml_node_t * node, int index) {
+    char * n;
+    char * v = mxmlElementGetAttrByIndex(node, index, &n);
+    if (v) {
+        return bbStringFromUTF8String(v);
+    } else {
+        return &bbEmptyString;
+    }
+}
+
+BBString * bmx_mxmlElementGetAttrCaseInsensitive(mxml_node_t *node,
+                                                   BBString *attr,
+                                                   int *found) {
+    if (attr == &bbEmptyString) return &bbEmptyString;
+
+    char *a = bbStringToUTF8String(attr);
+    const char *result = NULL;
+    int type = mxmlGetType(node);
+    
+    *found = 0;
+
+    if (type == MXML_ELEMENT && a) {
+        int count = mxmlElementGetAttrCount(node);
+
+        for (int i = 0; i < count; i++) {
+            const char *name = NULL;
+            const char *val = mxmlElementGetAttrByIndex(node, i, &name);
+
+            if (!name) 
+                continue;
+
+            // different name?
+            if (strcasecmp(name, a) != 0) {
+                continue;
+            }
+
+            // found the value
+            result = val;   // can be NULL !
+            *found = 1;
+            break;
+        }
+    }
+
+    bbMemFree(a);
+
+    return bbStringFromUTF8String(result);
```

Comment:
Who frees result?

Reply:
which one? result? bmx_mxmlElementGetAttr did not do that neither.

Reply:
`result` is not allocated here, so there is nothing to free.

`mxmlElementGetAttrByIndex()` returns a pointer to an internally managed string owned by mxml. Its lifetime is bound to the `mxml_node_t`, and it must not be freed by the caller.

The only allocation in this function is:
```
char *a = bbStringToUTF8String(attr);
```

which is correctly released with:
```
bbMemFree(a);
```

The return value:
```
return bbStringFromUTF8String(result);
```

creates a new `BBString` by copying the contents of result, so the returned string is independent of the XML node’s lifetime.


Edit: mxml definition of the index function:
```
const char *mxmlElementGetAttrByIndex(mxml_node_t *node, size_t idx, const char **name);
```